Feat/opt user management

akomii opened the pull request:
Added local authentication multi-user module with optional YubiKey OTP support for the AKTIN Broker server. 

Handles username/password login, first-time OTP registration, and token verification, with secure logging and configurable enforcement.

README and Docs will be added after accepted PR

akomii committed:
copied cred package for modification

akomii committed:
changed broker-admin paths and packages to cred2

akomii committed:
added interface for database changelog hooks

akomii committed:
updated LiquibaseWrapper to run custom changelogs, added getter to CascadedAuthProvider

akomii committed:
added custom db changelog to CredentialTokenAuthProvider

akomii committed:
added feature to initialiseDatabase to run custom changelogs after db creation

akomii committed:
Only Formatting

akomii committed:
added more columns to users table

akomii committed:
refactored token, added userId, moved changelog filepath into var

akomii committed:
changed user primary key to username

akomii committed:
added user pojo and userDao interface

akomii committed:
aligned User Pojo to DB schema

akomii committed:
added implementation for UserDao

akomii committed:
smaller fixes, removed toString() from User (for now)

akomii committed:
refactored token principal
- implemented proper random GUID generation
- implemented renew() and invalidate()
- added configurable token lifespan

akomii committed:
broke up classes into submodules

akomii committed:
updated visibility of TokenManager and UserJdbcRepository, added password hasher + implementation

akomii committed:
added update operation to UserRepository

akomii committed:
added user service and implementation

akomii committed:
added more logging to JdbcUserService, added path for unchanged default password

akomii committed:
smaller fixes / renaming for consistency

akomii committed:
added user auth service and implementation, added ToDo

akomii committed:
changed TokenManager to an interface, added implementation,

akomii committed:
made JdbcUserRepository a Singleton, changed renewing of token to sliding-window

akomii committed:
added periodic token cleanup service

akomii committed:
refactored CredentialTokenAuth to properly check and renew tokens

akomii committed:
minor fixes

akomii committed:
refactored AuthEndpoint

akomii committed:
refactored UserEndpoint, added UserDTO

akomii committed:
fixes

akomii committed:
renamed db column "alg" to "algorithm" for clarity

akomii committed:
updated docs, renamed impl of UserAuthService (not dependent on jdbc)

akomii committed:
added proper catching of exceptions to endpoint

akomii committed:
minor fixes

akomii committed:
changed visibility of singletons, updated constructor of JdbcUserService to not throw exceptions

akomii committed:
refactored CredentialTokenAuthProvider

akomii committed:
added custom intialization form DataSource in JdbcUserService

akomii committed:
fixed duplicate unique constraint in db, switched TestServer back to using cred (instead of cred2)

akomii committed:
added TestConfiguration for cred2

akomii committed:
added proper logging of exception to UserEndpoint

akomii committed:
FIX: wrong column in mapper

akomii committed:
added lazy initialization of default user

akomii committed:
change type of createdAt from Instant to long

akomii committed:
added endpoint constraints
- default user cannot be deactivated
- duplicate username cannot exist

akomii committed:
added ToDos / minor fixes

akomii committed:
recerted default auth provider to cred

akomii committed:
reverted all changes outside cred2

akomii committed:
moved cred2 components into own module, removed TokenPruneService

akomii committed:
added broker-auth-multiuser-otp to build order of parent pom, removed broker-auth-openid from broker-admin-dist

akomii committed:
renamed http to controller

akomii committed:
renamed auth to token

akomii committed:
appended DTO to controller pojo names

akomii committed:
added some ToDos

akomii committed:
remnoved JDBC implementation and userCreds.xml

akomii committed:
added implementation for CRUD operations on filesystem, updated UserRepository to return boolean instead of void with exceptions

akomii committed:
added filesystem impl for UserService, updated UserService to return boolean instead of void with exceptions

akomii committed:
streamlined logging messages

akomii committed:
refactored endpoints to better log user actions, outsourced some logic into utils class

akomii committed:
adjusted token auth provider

akomii committed:
fixed brittle test

akomii committed:
moved back in: broker-auth-opend (app doesnt work correctly without it???)

akomii committed:
FIX: Cred2TestConfiguration WIP, credentialsDTO was missing root name

akomii committed:
removed update feature of default user

akomii committed:
added optional field for otp token to user pojo, adjusted implementation

akomii committed:
updated UserService to set otp tokens

akomii committed:
smaller fixes, added interface for otp verification

akomii committed:
added optional verification of otp tokens

akomii committed:
simplified repository functions

akomii committed:
improved auditing in endpoint

akomii committed:
init otp verification service implementation

akomii committed:
rechecked logging and logic in YubicoVerificationService

akomii committed:
added check for public id

akomii committed:
added otp field to CredentialsDTO, added binding of otp classes to CredentialTokenAuthProvider

akomii committed:
added var to UserDTO to check if user has otp token

akomii committed:
FIX: opt check order

akomii committed:
updated desc of pom.xml

akomii committed:
added optional input field for tokens in UI

akomii committed:
removed Cred2TestConfiguration.java

rwm commented on broker-admin/src/main/resources/webapp/login.html:
```diff
@@ -27,9 +27,10 @@
+        var xml = $.parseXML("<credentials><username/><password/><token/></credentials>");
         $(xml).find('username').text($('#user').val());
         $(xml).find('password').text($('#pass').val());
+        $(xml).find('token').text($('#otp').val());
```
otp should not be mentioned here, as it is too specific. we should stick with the generic "token"

akomii replied:
I chose "otp" because "token" is duplicate with a var name in LOC40: 

https://github.com/aktin/broker/blob/69079bf47565e1e2de55f15cf90f2070266e897c/broker-admin/src/main/resources/webapp/login.html#L39-L42

akomii replied:
Resolved in 8aa24bbac2e2819d1926fa99219455cc20ba587b

rwm commented on broker-admin/src/main/resources/webapp/login.html:
```diff
@@ -55,6 +56,7 @@ <h1>Broker Server</h1>
 			<form id="login">
 				<label>Login <input type="text" id="user"/></label><br/>
 				<label>Password <input type="password" id="pass"/></label><br/>
+				<label>Token (optional) <input type="text" id="otp"/></label><br/>
```
no "otp" here, see above. And also remove "optional" as this is application dependent.

akomii replied:
Resolved in 8aa24bbac2e2819d1926fa99219455cc20ba587b

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/controller/UserEndpoint.java:
```diff
@@ -0,0 +1,116 @@
+  @POST
+  @Authenticated
+  @RequireAdmin
+  @Path("{username}/activate")
```
This is not clean REST, but sounds more like HTTP RPC. REST methods should not sound like method calls.
With minimal effort, consider something like 
POST /projects/:id/activation       <-- activate project
DELETE /projects/:id/activation     <-- deleting the activation = deactivate

akomii replied:
Resolved in dae8b04d3af757a463e1114d051b6a2294ca59bc

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/controller/UserEndpoint.java:
```diff
@@ -0,0 +1,116 @@
+  @Consumes(MediaType.APPLICATION_XML)
+  public Response create(@HeaderParam(HttpHeaders.AUTHORIZATION) String bearer, CredentialsDTO cred) {
+    Token token = EndpointUtils.requireDefaultAdmin(bearer, manager);
+    log.info(String.format("User %s attempts to create new user: %s", token.getName(), cred.username));
```
Logs should indicate what happened, not what might happen. We dont want multiple info logs for creating a single user. Please merge this message with the messages below.

akomii replied:
Resolved in eb79d12b5c9514f762b63e2a05cde22dc844778a

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/repository/FsUserRepository.java:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  private void loadUsersIntoCache() {
+    lock.writeLock().lock();
```
The locking strategy seems messed up. reading sometimes uses writeLock (loadUsersIntoCache) and sometimes uses readLock (find) and somethimes not locks at all e.g. saveUsersToFile. Also uses ConcurrentHashMap which already handles concurrency for reading so no locking should be needed for accessing the map. loadUsersIntoCache is called in constructor and therefore does not need writeLocking because this cannot coincide with any other call which needs to come after construction. please revise/simplify or discuss with me.

akomii replied:
- `loadUsersIntoCache()` is using a writeLock because the state of `userCache` is modified. I want to prohibit all read operations to shared data. As its called in constructor, it the writeLock could be dropped
- `saveUsersToFile()` is only used within `insert()` and `update()` which both have the writeLock while calling the method (needs to be documented better)
- Yes, `ConcurrentHashMap` is redundant and should be replaced with a regular `HashMap`

akomii replied:
Resolved in 05d88a541dd06fe608b8a9fb7c03fc8f3122fc6b

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/service/FsUserService.java:
```diff
@@ -0,0 +1,95 @@
+      String raw;
+      if (password == null || password.isBlank()) {
+        raw = generateRandomPassword();
+        log.info(String.format("Creating default user %s with random password %s", username, raw));
```
Is this useful? During the first start, it will output the random password in the log. After second start, nobody knows the password anymore. The user cannot be expected to read info log messages and write down anything from this. Maybe rather remove this code and also remove the random password generation.

akomii replied:
Random Password generation was a feature migrated from the old authenticator because of backwards-compability (feature-wise). Will be removed then.

akomii replied:
If the password is required as a system property, how should it be handled if the password changes?

akomii replied:
Extern Call : Password should be set after every startup from system.properties

akomii replied:
Resovled in aa91a7eb8a12233e2197516d7334102b65d12f16

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/service/UserAuthServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+  @Inject
+  public UserAuthServiceImpl(UserService service, PasswordHasher hasher, OtpVerificationService otpVerificationService) {
+    this.userService = Objects.requireNonNull(service);
+    this.passwordHasher = Objects.requireNonNull(hasher);
```
Password hasher should not be needed to be used in multiple classes (user auth, repository). This increases complexity in class-class dependencies. Instead, use passwordHasher only in userService. then passwordhasher does not need any dependency injection.

akomii replied:
Resolved in a3c533174f3649e4b634fce0be4efd4a9c1689a6

akomii replied:
I am not very  happy with changes. While `PasswordHasher` is now only used in `FsUserService`, `FsUserService` now partly does authentication (which I would assume would only happen in `UserAuthServiceImpl`). Also, there is now a style inconsistency in `UserService` as I used `User` as an input argument (instead of `username`) to avoid refetching the object:

> boolean verifyUserPassword(User user, char[] password);

Do you know of a better way to solve this issue?

akomii replied:
Extern Call : Will be kept as is.

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/CredentialTokenAuthProvider.java:
```diff
@@ -0,0 +1,64 @@
+  public void bindSingletons(BiConsumer<Object, Class<?>> binder) {
+    binder.accept(manager, TokenManager.class);
+    binder.accept(repository, UserRepository.class);
+    binder.accept(hasher, PasswordHasher.class);
```
The hasher should not be a singleton and does not need to be used for injection. See comment below in UserAuthServiceImpl.

akomii replied:
Resolved in a3c533174f3649e4b634fce0be4efd4a9c1689a6

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/token/Token.java:
```diff
@@ -0,0 +1,97 @@
+  private static final long DEFAULT_TTL_SECONDS = 360L;
+
+  private static final int ID_BYTES = 32;
+  private static final SecureRandom RANDOM = new SecureRandom();
```
static final securerandom is bad practice, because it is not reproducible e.g. for testing and cannot involve seed.
The generation of tokenids is not concern of the token but instead should happen in TokenManager. There, the secure random can be a member variable.

akomii replied:
Resolved in 800b092d2c0954e6f8c7f605003b342a543432e7

akomii replied:
Token has now no checks for `tokenTimeToLive` and has to trust `TokenManagerImpl`. I dont see any problem with that. What do you think?

akomii replied:
Extern Call: Fine as is

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/controller/AuthEndpoint.java:
```diff
@@ -0,0 +1,75 @@
+  @Produces(MediaType.APPLICATION_XML)
+  public StatusDTO status(@HeaderParam(HttpHeaders.AUTHORIZATION) String bearer) {
+    Token t = EndpointUtils.resolveTokenFromBearerHeader(bearer, manager);
+    StatusDTO s = new StatusDTO();
```
Simpler to use return new StatusDTO and put @AllArgsConstructor there

akomii replied:
Seems to me more like a styling decision... Adding Lombok just to simplify `StatusDTO` doesn’t justify the extra dependency. If Lombok is added, the annotations should be applied consistently across all POJOs, which would mean a larger refactoring effort....

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/service/FsUserService.java:
```diff
@@ -0,0 +1,95 @@
+
+  @Override
+  public OperationResult setToken(String username, String token) {
+    if (token == null || token.length() < 12) {
```
This hardwires yubikeyotp in here. this breaks any abstraction/interface layer for OTP. We can keep this here, but then then remove any interfaces for OTP and stick with hardwired for Yubikey. Alternatively, rethink what/how the fsservice should know about the OTP method.

akomii replied:
Resolved in a3c533174f3649e4b634fce0be4efd4a9c1689a6

akomii replied:
For more clarity, would it make sense to rename in `User`

>   public final Optional<String> tokenProvider;
>   public final Optional<String> token;

to

>   public final Optional<String> otpProvider;
>   public final Optional<String> binding;

akomii replied:
Extern Call: OtpProvider and algorithm should be removed from persistence

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/repository/FsUserRepository.java:
```diff
@@ -0,0 +1,188 @@
+          User user = parseUser(parts);
+          userCache.put(user.username, user);
+          loadedCount++;
+        }
```
What about parts.length < 5? any invalid/ignored lines should at least log a warning.

rwm replied:
Also we should expect exactly 6 parts with the token part being empty sometimes

akomii replied:
Resolved in 05d88a541dd06fe608b8a9fb7c03fc8f3122fc6b

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/repository/FsUserRepository.java:
```diff
@@ -0,0 +1,188 @@
+
+  private static final String PROPERTY_USER_FILE = "aktin.broker.users.file";
+  private static final String DEFAULT_USER_FILE = "users.txt";
+  private static final String FIELD_SEPARATOR = ";";
```
Dont use ; instead use 'tab' because this is more unlikely to occur inside a password or the hash algorithm. Otherwise this will be a security issue due to the lack of escaping.

akomii replied:
Resolved in 05d88a541dd06fe608b8a9fb7c03fc8f3122fc6b

rwm reviewed (changes requested):
please adress the comments/changes and resubmit or discuss with me before merging.

akomii committed:
renamed var "otp" to more generic "token"

akomii committed:
updated user activation/deactivation endpoints to be more RESTful

akomii committed:
log actual authentication and user management outcomes

akomii committed:
Refactored FsUserRepository according to code review:
- added readwritelock as input to saveUsersToFile() to resemble dependency to insert() and update()
- changed field separator to tabs to avoid symbol occurence in passwords
- removed concurrency from loadUsersIntoCache() as only used in constructor
- removed wait for cache load
- changed user load logic to expect always 6 parts in line of users file
- added warning for malformed lines in initial load

akomii committed:
moved guid generation and life span setting from Token to TokenManagerImpl

akomii committed:
moved passwordhasher from UserAuthServiceImpl to FsUserService
- updated UserService to fulfill passwordhasher methodology (algorithm check, password verification)
- replaced passwordhasher calls in UserAuthServiceImpl with FsUserService
- removed singleton binds of passwordhasher and otpverification service in CredentialTokenAuthProvider

akomii committed:
cleaned up hardwiring of yubico otp
- renamed OtpVerificationService to OtpProvider
- added methods to get binding and id to otpProvider interface
- added otp provider var to user pojo
- updated persistence logic for new otp provider var
- updated UserDTO to keep otp provider instead of boolean if otp is set
- added otp verification methods to UserService
- moved otp provider dependency from UserAuthServiceImpl to FsUserService

akomii committed:
removed case differentiation from default user init

akomii committed:
removed used hash algorithm and otpProvider from User Pojo and persistence

akomii committed:
added README

akomii commented:
Required changes are done. Docs and Tests are still missing. Where should the tests go? Into `dwh-admin`?

rwm commented on broker-auth-multiuser-otp/README.md:
```diff
@@ -0,0 +1,27 @@
+
+## Usage
+1. Build this package
+2. Move the created `.jar` to the other `.jars` of the AKTIN Broker (by default in the `lib/` folder)
```
.jars ist keine Endung : )

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/repository/FsUserRepository.java:
```diff
@@ -0,0 +1,188 @@
+        Files.createFile(usersFile);
+        log.info("Created users file: " + usersFile.toAbsolutePath());
+      } catch (IOException e) {
+        log.severe("Could not create users file: " + e.getMessage());
```
exception geloggt und danach alles gut? vermutlich kann die anwendung nicht weiter laufen.. dann braucht man die exception auch nicht abfangen sondern müsste weiter oben abgefangen werden oder zum absturz führen.
Kritische Fehler abfangen und dann nur Lognachricht grundsätzlich vermeiden.

rwm commented on broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/repository/FsUserRepository.java:
```diff
@@ -0,0 +1,174 @@
+
+  private final Path usersFile;
+  private final Map<String, User> userCache = new HashMap<>();
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
```
Hier bitte doku ergänzen, dass das lock gebraucht wird damit die Datei nicht parallel geschrieben und zerstört wird.
Man denkt sonst du machst das für die Hashmap, was unnötig wäre. Evtl auch unten kurz kommentieren wo das writeLock benutzt wird.

rwm replied:
ist immernoch ein bisschen messy, weil du das lock gleichzeitig für die hashmap und den filezugriff benutzt..
z.B. readLock zum lesen der hashmap. aber dann behältst du nach dem update der hashmap das writelock während die datei geschrieben wird, was formell falsch ist (weil da die hashmap nur gelesen werden muss).
Simplere alternative: ConcurrentHashMap für den zugriff auf die map und ReentrantLock(true) für die Schreiboperation der Datei. 
Kannst du auch so lassen und nur Kommentare ergänzen - ist nur ausführliches Feedback von mir weil etwas verwirrend beim lesen/verstehen ohne doku.

akomii replied:
Da ist nichts formell falsch. Es ist einfach nur ein anderer Fokus. Hier gehts doch gerade um simplicity/clarity vs. granularity/correctness. Bei einer ConcurrentHashMap kann es ja auch passieren, dass A was reinschreibt, B genau das liest, aber der Server abstürzt bevor es persistert werden kann. Ich persönlich finde es einfacher, wenn die HashMap ein geschlossener Raum ist, der nur von einer Person betreten werden darf, ohne Schaufenster o.ä.

rwm reviewed (approved):
please read and address comments before merging.

akomii committed:
fixed typo in README

akomii committed:
added docs to controller package

akomii committed:
added docs to service package

akomii committed:
added docs to token package

akomii committed:
added docs to utils package and CredentialTokenAuthProvider

akomii committed:
added docs to repository classes

akomii committed:
updated FsUserRepository
- added EXPECTED_FIELD_COUNT to avoid magic number
- init lock as ReentrantReadWriteLock to avoid casting
- added IllegalStateException if initial file load fails
- added RuntimeException if write operation files

akomii committed:
added docs to FsUserRepository
